src/proofs/Term.cpp

```diff
@@ -0,0 +1,720 @@
+/*
+ * Copyright (c) 2023, Matias Barandiaran <matias.barandiaran03@gmail.com>
+ *
+ * SPDX-License-Identifier: MIT
+ */
+
+#include "Term.h"
+#include "FastRational.h"
+#include <cassert>
+#include <memory>
+#include <string>
+
+std::string Term::printTerm() {
+    PrintVisitor printVisitor;
+    this->accept(&printVisitor);
+    return printVisitor.getString();
+}
+
+void PrintVisitor::visit(Terminal * term) {
+    ss << term->getVal();
+}
+
+void PrintVisitor::visit(Op * term) {
+    ss << "(" << term->getOp();
+    for (const std::shared_ptr<Term> & arg : term->getArgs()) {
+        ss << " ";
+        arg->accept(this);
+    }
+    ss << ")";
+}
+
+void PrintVisitor::visit(App * term) {
+    ss << "(" << term->getFun();
+    for (const std::shared_ptr<Term> & arg : term->getArgs()) {
+        ss << " ";
+        arg->accept(this);
+    }
+    ss << ")";
+}
+
+void PrintVisitor::visit(Quant * term) {
+    ss << "(" << term->getQuant() << " (";
+    for (int i = 0; i < term->getVars().size(); i++) {
+        ss << "(";
+        term->getVars()[i]->accept(this);
+        ss <<" ";
+        term->getSorts()[i]->accept(this);
+        ss << ")";
+        if (i + 1 != term->getVars().size()) { ss << " "; }
+    }
+    ss << ") ";
+    term->getCoreTerm()->accept(this);
+    ss << ")";
+}
+
+void PrintVisitor::visit(Let * term) {
+    auto names = term->getTermNames();
+    ss << "(let (";
+    for (int i = 0; i < names.size(); i++) {
+        ss << "(" << names[i] << " ";
+        term->getDeclarations()[i]->accept(this);
+        ss << ")";
+        if (i + 1 != names.size()) { ss << " "; }
+    }
+    ss << ") ";
+    term->getApplication()->accept(this);
+    ss << ")";
+}
+
+std::string Op::simplifyRule() {
+    std::string op = operation;
+    if (op == "=") {
+        if ((args[0]->printTerm().find_first_not_of("( )-0123456789") == std::string::npos) and
+            (args[0]->printTerm().find_first_not_of("( )-0123456789") == std::string::npos)) {
+            return "eq_simplify";
+        } else {
+            return "equiv_simplify";
+        }
+    } else if ((op == ">") or (op == "<") or (op == "<=") or (op == ">=")) {
+        return "comp_simplify";
+    } else if (op == "and") {
+        return "and_simplify";
+    } else if (op == "or") {
+        return "or_simplify";
+    } else if (op == "+") {
+        return "sum_simplify";
+    } else if (op == "-") {
+        return "minus_simplify";
+    } else if (op == "/" or op == "div") {
+        return "div_simplify";
+    } else if (op == "*") {
+        return "prod_simplify";
+    } else if (op == "not") {
+        return "not_simplify";
+    } else if (op == "ite") {
+        return "ite_simplify";
+    } else if (op == "mod") {
+        return "mod_simplify";
+    }
+    return "Error";
+}
+
+std::string NegatedAndVisitor::visit(Op * term) {
```

Comment:
It seems to me you do not need to implement this functionality in a Visitor, but a small helper function should be sufficient.

Reply:
This visitor handles non linearity, i.e:

(and true true P x Q y R z) 
-> (not Px) (not Qy) (not Rz)

Reply:
That's OK, I am just saying that we do not need a visitor for that, but the functionality can be implemented in a small helper function.
Because this visitor is not really visiting the whole term, it really is only an operation on the top level of an and-term.